### ISSUE TITLE
Disallow yyyy-dd-mm date format (and only allow 4-digit years)

### ISSUE THREAD

**Issue by PallHaraldsson** [JuliaLang/julia#52348]
Note, I didn't write, not to be confused with, yyyy-mm-dd.

A.
I didn't think this was an actual problem but saw at:

https://go.dev/doc/go1.20
>The vet tool now reports use of the time format 2006-02-01 (yyyy-dd-mm) with [Time.Format](https://go.dev/pkg/time/#Time.Format) and [time.Parse](https://go.dev/pkg/time/#Parse). This format does not appear in common date standards, but is frequently used by mistake when attempting to use the ISO 8601 date format (yyyy-mm-dd).

I see we allow both conflicting formats:
```
julia> DateTime("20150102",dateformat"yyyyddmm")
2015-02-01T00:00:00

julia> DateTime("20150102",dateformat"yyyymmdd")
2015-01-02T00:00:00
```

I don't think we should WARN/report, just error on this, since it serves no useful purpose. Preemptively stop this from happening in Julialand, it might not be common here (yet).

Since this is done with a dateformat-string, i.e. a macro, I think can be no cost, at compile time, when format is correct, erroring otherwise.

Since the legal format is yyyy-mm-dd with hyphens, should only that be allowed? At least not have the format without hyphens in the docs?

B.
This does work:
```
julia> DateTime("22015-01-02",dateformat"yyyy-mm-dd")
22015-01-02T00:00:00
```

but actually we should also disallow 5-digit years, since a possible typo and it's NOT allowed by the ISO spec. Probably since it could be common typo.

Non-4-digit are only allowed by preexisting arrangement, and then by prefixing with a + (or a -).

We should allow that however (but no pressure to define just yet), at least with a keyword opt-in, though I do not recall if the standard requires such an opt-in.

julia> DateTime("+222012-15-01-02")
ERROR: ArgumentError: Invalid DateTime string

julia> DateTime("0-15-01-02")
ERROR: ArgumentError: Invalid DateTime string

julia> DateTime("0-12-01")
0000-12-01T00:00:00

That year DOES NOT exist (and should be disallowed), one reason the + is required. AD 1 preceded 1 BC. But +0 (or -0 I guess) can be defined as AD 1, and -1 as AD 2 etc.

The only sort of good reason to allow 0 in year (as opposed to +0, or "0000") might be if it in some form, is used in some common "data" such as 0000-01-01 as an illegal date format for a missing value. I think actually 0000-00-00 or 9999- sometime might be used, and we could allows such 0 to be parsed as "missing" even, but I'm not posing for that, only to get it disallowed for now. Or at least year 1 (2?) to 1582.

TL;DR olly allow year range 1583 to 9999, not 5-, 6- etc years. Some standard (Ecma if I recall) allows higher than 4-digit but if I recall then only 6-digit with a plus, even for 5-digit year.

It is actually disallowed to use the year 1582 or older, by ISO 8601, what we follow, except "shall only be used by mutual agreement of the partners in information interchange." I think we should disallow parsing it, to open up possibility of new interpretation of two-digit YY years, same as JavaScript does.

It's at least not very helpful to allow YY years (differently to others) or pre-1583 year, possible typos.

People are sometimes strict on what they output, but lenient on what they input, so we could allow parsing with status qu, at least for a while, but I think it's better to stop doing that ASAP, since not very useful anyway, and allowed, even required to do. Then with + prefix can be added at same or later point, to allow that back.

The Gregorian calendar starts at 5 October 1582. while the ISO 8601 we use (or at least ISO/DIS 8601-1:2016(e)) allows more exceptionally, i.e.:

https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar

https://web.archive.org/web/20171019211402/https://www.loc.gov/standards/datetime/ISO_DIS 8601-1.pdf

ISO/DIS 8601-1:2016(e)

> This International Standard allows the identification of calendar years by their year number for years
> both before and after the introduction of the Gregorian calendar. For the determination of calendar
> years, the year number and the calendar day within the calendar year only the rules mentioned above
> are used. For the purposes of this International Standard the calendar based on these rules is referred to
> as the Gregorian calendar. The use of this calendar for dates preceding the introduction of the Gregorian
> calendar (also called the proleptic Gregorian calendar) should only be by agreement of the partners in
> information interchange.
> 
> The introduction of the Gregorian calendar included the cancellation of the accumulated inaccuracies of
> the Julian calendar. However, no dates shall be inserted or deleted when determining dates in the
> proleptic Gregorian calendar.
> 
> NOTE In the proleptic Gregorian calendar, the calendar year [0000] is a leap year.
> 
> EXAMPLE The Gregorian calendar was introduced on 15 October 1582. In the calendar set by this standard
> the calendar day preceding that calendar day is referred to as 14 October 1582. In the Julian calendar that
> calendar day is referred to as 4 October 1582.

> The representations specified in this International Standard make use of graphic characters as specified
> in 3.4. Note that, except for “hyphen”, “minus” and “plus-minus”, these characters are part of the ISO/IEC
> 646 repertoire.
>
> In an environment where use is made of a character repertoire based on ISO/IEC 646, “hyphen” and
> “minus” are both mapped onto “hyphen-minus”. Representations with a “plus-minus” shall only be used
> in such environment if the interchange repertoire includes “plus-minus”.

> By mutual agreement of the partners in information interchange, it is permitted to expand the
> component identifying the calendar year, which is otherwise limited to four digits. This enables
> reference to dates and times in calendar years outside the range supported by complete representations,
> i.e. before the start of the year [0000] or after the end of the year [9999].

> *calendar year* is, unless specified otherwise, represented by four digits. Calendar years are
> numbered in ascending order according to the Gregorian calendar by values in the range [0000]
> to [9999]. Values in the range [0000] through [1582] shall only be used by mutual agreement of
> the partners in information interchange.

**Comment by PallHaraldsson** [comment]
A.
I realize Go vet is a linter, does that mean we need to support arbitrary things in dateformat_str macro? People could always work around that not use the macro, if you want to support incorrect date formats. We should make correct use easy, and incorrect use hard(er) but it wouldn't still be impossible.

B.
I had two-in-one issue, and I think this should be open for B., I'm at least working on a PR for it. Unless you don't want to follow the standard, and not open, then it frees up my time to not work on that PR.

**Comment by mbauman** [comment]
It's always best to keep issues focused on one thing at a time. In any case, discussion on that part B can continue in your PR at #52351.
